sqlflow_models/dbscan.py

```diff
@@ -0,0 +1,179 @@
+    OUTLIER = -1
+
+    def __init__(self, min_samples=2, eps=10, feature_columns=None):
+        super(DBSCAN, self).__init__(name='DBSCAN')
+        self.minpts = min_samples
+        self.eps = eps
+        self.clusters = []
+        self.labels_ = []
+
+    def call(self):
+        pass
+
+    def _to_dataframe(self, dataset):
+        x_df = pd.DataFrame()
+        y_df = pd.DataFrame()
+
+        for features, label in dataset:
+            dx = {}
+            dy = {}
+            for name, value in features.items():
+                dx[name] = value.numpy()[0]
+            x_df = x_df.append(dx, ignore_index=True)
+            if label is not None:
+                dy['label'] = label.numpy()[0][0]
+                y_df = y_df.append(dy, ignore_index=True)
+
+        if y_df.empty:
+            return x_df, None
+        return x_df, y_df['label']
+
+
+
+    def intersect(self, a, b):
+        return len(list(set(a) & set(b))) > 0
+
+    def compute_neighbors(self, distance_matrix):
+        neighbors = []
+        for i in range(len(distance_matrix)):
+            neighbors_under_eps = []
+            for neighbor in range(len(distance_matrix[i])):
+                if distance_matrix[i][neighbor] <= self.eps \
+                        and neighbor != i:
+                    neighbors_under_eps.append(neighbor)
+            neighbors.append(neighbors_under_eps)
+        return neighbors
+
+    def generate_clusters(self, neighbors_list):
+        # initiate with the first data
+        clusters = [neighbors_list[0] + [0]]
+        for i in range(1, len(neighbors_list)):
+            # for other data in the neighbour list
+            # check if the data has an intersected cluster inside the result list
+            # merge the list and append it to the result
+            list_of_intersected_cluster = []
+            new_cluster = neighbors_list[i] + [i]
+            for cluster_num in range(len(clusters)):
+                if self.intersect(neighbors_list[i],
+                                  clusters[cluster_num]):
+                    list_of_intersected_cluster.append(clusters[cluster_num])
+                    new_cluster = new_cluster + clusters[cluster_num]
+
+            # if the data is a new cluster / no intersected clusters
+            if not list_of_intersected_cluster:
+                clusters.append(neighbors_list[i] + [i])
+            else:
+                clusters.append(list(set(new_cluster)))
+                # delete the merged clusters
+                for old_cluster in list_of_intersected_cluster:
+                    clusters.remove(old_cluster)
+        return clusters
+
+    def labelling(self, data, clusters):
+        cluster_labels = [self.OUTLIER] * len(data)
+        for i in range(len(self.clusters)):
+            for j in range(len(self.clusters[i])):
+                cluster_labels[self.clusters[i][j]] = i
+        return cluster_labels
+
+    def fit(self, X):
```

Comment:
Instead of overriding `fit` function, we can implement `sqlflow_train_loop` to define the custom train loop logic. You may find that `sqlflow_train_loop` is not the standard API of Keras model, it's just let SQLFlow `runtime` to know to run the custom train loop.

c.f. https://github.com/sql-machine-learning/sqlflow/blob/9628487783878e9f019b1bf379e192134344e3a3/python/sqlflow_submitter/tensorflow/train_keras.py#L173

Reply:
Enables `sqlflow_train_loop` function to handle tf.dataset type of data, because `sqlflow` server  invokes tf.dataset.
[https://github.com/sql-machine-learning/sqlflow/blob/483b8676cf93f373d5073d84b0bee311bb122012/python/runtime/tensorflow/input_fn.py#L72](url)